File: Python/DesignPatterns/DecoratorDesignPattern.py
from abc import ABC, abstractmethod


class Notification(ABC):
    @abstractmethod
    def send(self, message: str) -> None:
        raise NotImplementedError


class SMSNotification(Notification):
    def send(self, message: str) -> None:
        print("Sending SMS in a legacy way")


class WhatsappNotification(Notification):
    def send(self, message: str) -> None:
        print("Sending SMS in a legacy way")


# Retry // Formatting
class NotificationDecorator(Notification, ABC):
    def __init__(self, notification: Notification) -> None:
        self.wrapped_notification = notification


class RetryDecorator(NotificationDecorator):
    def __init__(self, notification: Notification) -> None:
        super().__init__(notification)

    def send(self, message: str) -> None:
        print("We are retrying with lots of efforts")
        self.wrapped_notification.send(message)


class FormattingDecorator(NotificationDecorator):
    def __init__(self, notification: Notification) -> None:
        super().__init__(notification)

    def send(self, message: str) -> None:
        print("Formatting code is running and processing")
        self.wrapped_notification.send(message)


class App:
    def __init__(self) -> None:
        self.notification: Notification | None = None

    def set_notification(self, notification: Notification) -> None:
        self.notification = notification

    def send_notification(self, message: str) -> None:
        self.notification.send(message)


def main() -> None:
    # notification = SMSNotification()  # can be taken care using factory design pattern
    # notification = FormattingDecorator(RetryDecorator(SMSNotification()))

    sms = SMSNotification()
    decorator1 = RetryDecorator(sms)
    formatting = FormattingDecorator(decorator1)

    app = App()
    app.set_notification(formatting)
    message = "Hello"
    app.send_notification(message)


if __name__ == "__main__":
    main()

File: Java/DesignPatterns/DecoratorDesignPattern.java

interface Notification{
    void send(String message);
}

class SMSNotification implements Notification{
    @Override
    public void send(String message) {
        System.out.println("Sending SMS in a legacy way");
    }
}

class WhatsappNotification implements Notification{
    @Override
    public void send(String message) {
        System.out.println("Sending SMS in a legacy way");
    }
}


// Retry // Formatting

abstract class NotificationDecorator implements Notification{
    protected Notification wrappedNotification;
    NotificationDecorator(Notification notification){
        this.wrappedNotification = notification;
    }
}

class RetryDecorator extends NotificationDecorator{

    public RetryDecorator(Notification notification){
        super(notification);
    }
    @Override
    public void send(String message) {
        System.out.println("We are retrying with lots of efforts");

        wrappedNotification.send(message);
    }
    
}

class FormattingDecorator extends NotificationDecorator{
   
    public FormattingDecorator(Notification notification){
        super(notification);
    }
    @Override
    public void send(String message) {
        System.out.println("Formatting code is running and processing");
        wrappedNotification.send(message);
    }
}


class App{

    Notification notification;
    void setNotification(Notification notification){
        this.notification = notification;
    }
    void sendNotification(String message){
        notification.send(message);
    }

}

public class DecoratorDesignPattern {
    public static void main(String[] args) {
        // Notification notification = new SMSNotification(); //can be taken care using factory design pattern 
        // Notification notification =    new FormattingDecorator(new RetryDecorator(new SMSNotification()));
        
        Notification SMS = new SMSNotification();
        Notification decorator1 = new RetryDecorator(SMS);
        Notification formatting = new FormattingDecorator(decorator1);
        
        App app = new App();
        app.setNotification(formatting);
        app.sendNotification(message);

        
    }    
}
